addons/addon-base-raas/packages/base-raas-services/lib/data-source/data-source-reachability-service.js

```diff
@@ -0,0 +1,197 @@
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+// const _ = require('lodash');
+const Service = require('@aws-ee/base-services-container/lib/service');
+const { allowIfActive, allowIfAdmin } = require('@aws-ee/base-services/lib/authorization/authorization-utils');
+
+const workflowIds = {
+  bulkCheck: 'wf-bulk-reachability-check',
+  accountStatusChange: 'wf-ds-account-status-change',
+};
+
+class DataSourceReachabilityService extends Service {
+  constructor() {
+    super();
+    this.dependency([
+      'auditWriterService',
+      'dataSourceAccountService',
+      'studyService',
+      'workflowTriggerService',
+      'aws',
+    ]);
+  }
+
+  async bulkReach(requestContext, { status }, forceCheckAll = false) {
+    status = status || '*';
+    const workflowTriggerService = await this.service('workflowTriggerService');
+    await workflowTriggerService.triggerWorkflow(
+      requestContext,
+      { workflowId: workflowIds.bulkCheck },
+      {
+        status,
+        forceCheckAll,
+      },
+    );
+    // Write audit event
+    await this.audit(requestContext, {
+      action: 'bulk-check-reachability',
+      body: { status },
+    });
+  }
+
+  async reachDsAccount(requestContext, { id, type }, forceCheck = false) {
+    const accountService = await this.service('dataSourceAccountService');
+    const dataSourceAccount = await accountService.mustFind(requestContext, { id });
+
+    if (!dataSourceAccount.status) {
+      throw this.boom.badRequest('Can only check reachability for data source account', true);
+    }
+    const prevStatus = dataSourceAccount.status;
+    let newStatus;
+    let statusMsg;
+
+    // TODO: Use the cloudformation stack name to access its status in the data source account
+    // TODO: Determine if the stack is compliant
+    const reachable = false;
+
+    if (reachable) {
+      newStatus = 'reachable';
+      statusMsg = '';
+    } else if (prevStatus === 'pending') {
+      newStatus = prevStatus;
+      statusMsg = `WARN|||Data source account ${id} is not reachable yet`;
+    } else {
+      newStatus = 'error';
+      statusMsg = `ERR|||Error getting information from data source account ${id}`;
+    }
+    if (prevStatus !== newStatus || forceCheck) {
+      const workflowTriggerService = await this.service('workflowTriggerService');
+      await workflowTriggerService.triggerWorkflow(
+        requestContext,
+        { workflowId: workflowIds.accountStatusChange },
+        {
+          id,
+          type,
+        },
+      );
+    }
+    await accountService.updateStatus(requestContext, dataSourceAccount, { status: newStatus, statusMsg });
+    const outputVal = newStatus;
+
+    // Write audit event
+    await this.audit(requestContext, {
+      action: 'check-dsAccount-reachability',
+      body: { id, type },
+    });
+
+    return outputVal;
```

Comment:
How about we return an object rather than just a string?  something like
{ status, statusMsg }
This will allow us in the future to also return the workflow instance id, which will help the caller of the API to check on the workflow status later.

Reply:
Done